Mathlib/NumberTheory/Cyclotomic/Rat.lean

```diff
@@ -331,16 +332,104 @@ theorem zeta_sub_one_prime [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+  simpa only [subOneIntegralPowerBasis_gen] using hζ.zeta_sub_one_prime
 
 theorem subOneIntegralPowerBasis'_gen_prime [IsCyclotomicExtension {p} ℚ K]
     (hζ : IsPrimitiveRoot ζ ↑p) :
-    Prime hζ.subOneIntegralPowerBasis'.gen := by simpa using hζ.zeta_sub_one_prime'
+    Prime hζ.subOneIntegralPowerBasis'.gen := by
+  simpa only [subOneIntegralPowerBasis'_gen] using hζ.zeta_sub_one_prime'
+
+open nonZeroDivisors in
+/-- The norm, relative to `ℤ`, of `ζ ^ p ^ s - 1` in a `p ^ (k + 1)`-th cyclotomic extension of `ℚ`
+is p ^ p ^ s` if `s ≤ k` and `p ^ (k - s + 1) ≠ 2`. -/
+lemma norm_toInteger_pow_sub_one_of_prime_pow_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) {s : ℕ} (hs : s ≤ k) (htwo : p ^ (k - s + 1) ≠ 2) :
+    Algebra.norm ℤ (hζ.toInteger ^ (p : ℕ) ^ s - 1) = p ^ (p : ℕ) ^ s := by
+  apply (algebraMap ℤ ℚ).injective_int
+  have : NumberField K := IsCyclotomicExtension.numberField {p ^ (k + 1)} ℚ K
+  have : algebraMap (𝓞 K) K (hζ.toInteger ^ (p : ℕ) ^ s - 1) = ζ ^ (p : ℕ) ^ s - 1 := by
+    simp only [map_sub, map_pow, map_one, sub_left_inj, RingOfIntegers.map_mk]
+  rw [← Algebra.norm_localization (Sₘ := K) ℤ ℤ⁰, this,
+    hζ.norm_pow_sub_one_of_prime_pow_ne_two
+      (cyclotomic.irreducible_rat (by simp only [PNat.pow_coe, gt_iff_lt, PNat.pos, pow_pos]))
+      hs htwo]
+  simp only [algebraMap_int_eq, map_pow, map_natCast]
+
+open nonZeroDivisors in
+/-- The norm, relative to `ℤ`, of `ζ ^ 2 ^ k - 1` in a `2 ^ (k + 1)`-th cyclotomic extension of `ℚ`
+is `(-2) ^ 2 ^ k`. -/
+lemma norm_toInteger_pow_sub_one_of_two [IsCyclotomicExtension {2 ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑((2 : ℕ+) ^ (k + 1))) :
+    Algebra.norm ℤ (hζ.toInteger ^ 2 ^ k - 1) = (-2) ^ (2 : ℕ) ^ k := by
+  apply (algebraMap ℤ ℚ).injective_int
+  have : NumberField K := IsCyclotomicExtension.numberField {2 ^ (k + 1)} ℚ K
+  have : algebraMap (𝓞 K) K (hζ.toInteger ^ 2 ^ k - 1) = ζ ^ (2 : ℕ) ^ k - 1 := by
+    simp only [map_sub, map_pow, map_one, sub_left_inj, RingOfIntegers.map_mk]
+  rw [← Algebra.norm_localization (Sₘ := K) ℤ ℤ⁰, this,
+    hζ.norm_pow_sub_one_two (cyclotomic.irreducible_rat (pow_pos (by decide) _))]
+  simp only [algebraMap_int_eq, Int.reduceNeg, map_pow, map_neg, map_ofNat]
+
+/-- The norm, relative to `ℤ`, of `ζ ^ p ^ s - 1` in a `p ^ (k + 1)`-th cyclotomic extension of `ℚ`
+is `p ^ p ^ s` if `s ≤ k` and `p ≠ 2`. -/
+lemma norm_toInteger_pow_sub_one_of_prime_ne_two [IsCyclotomicExtension {p ^ (k + 1)} ℚ K]
+    (hζ : IsPrimitiveRoot ζ ↑(p ^ (k + 1))) {s : ℕ} (hs : s ≤ k) (hodd : p ≠ 2) :
+    Algebra.norm ℤ (hζ.toInteger ^ (p : ℕ) ^ s - 1) = p ^ (p : ℕ) ^ s := by
+  have coe_two : ((2 : ℕ+) : ℕ) = 2 := by norm_cast
+  refine hζ.norm_toInteger_pow_sub_one_of_prime_pow_ne_two hs (fun h ↦ ?_)
+  rw [← PNat.coe_inj, coe_two, PNat.pow_coe, ← pow_one 2] at h
+  apply hodd
+  simp [← PNat.coe_inj, eq_of_prime_pow_eq hp.out.prime Nat.prime_two.prime (k - s).succ_pos h]
```

Comment:
I tried to use more general tactics so we wouldn't have to manually rewrite all the coercions. It's not a great golf but at least it's a bit less thinking.
```suggestion
  refine hζ.norm_toInteger_pow_sub_one_of_prime_pow_ne_two hs (fun h ↦ hodd ?_)
  suffices h : (p : ℕ) = 2 from PNat.coe_injective h
  apply eq_of_prime_pow_eq hp.out.prime Nat.prime_two.prime (k - s).succ_pos
  rw [pow_one]
  exact_mod_cast congr_arg Subtype.val h
```